### PR TITLE
fix(status): fetch volume status using controller podIP

### PR DESCRIPTION
**Pull request by shubham14bajpai**
Signed-off-by: shubham <shubham.bajpai@mayadata.io>

This PR address 2 issues:
 - Rapidly cycling Unknown and Ready status:
     Reason: Fetching the stats using the service IP repeatedly can cause dns issues which fails occasionally and causes cycling of status
     Fix: Use controller pod IP to fetch the stats and maintain an updated map for all volumes and pod IPs 
 - Log flooding of failed tcp requests
      The above fix will help here as well

Tests covered manually:
 - volume creation and check status
 - multiple controller pod restarts 
 - multiple replica pod restarts

### CODE REVIEW

**Comment by payes** [comment]
Can you please verify the following scenarios where a node containing both controller and replica is brought down:
1. Delete the controller pod, so that now 2 controller pods are present. One in deleting state and the other in Running/ContainerCreating state.
2. Verify if we are able to fetch the status even when one replica is down.

File: pkg/controllers/jivavolume_controller.go [hunk 1]
```diff
@@ -135,6 +136,16 @@ func (r *JivaVolumeReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 		return reconcile.Result{}, err
 	}
 
+	if _, ok := podIPMap[instance.Name]; !ok {
+		err = r.updatePodIPMap(instance)
+		if err != nil {
+			// log err only, as controller must be in container creating state
+			// don't return err as it will dump stack trace unneccesary
+			logrus.Infof("failed to get controller pod ip for volume %s: %s", instance.Name, err.Error())
+			time.Sleep(1 * time.Second)
```
Comment (prateekpandey14) [inline pkg/controllers/jivavolume_controller.go:145]: sleep required ? i get the idea but do we need to have it ?

File: pkg/controllers/jivavolume_controller.go [hunk 2]
```diff
@@ -135,6 +136,16 @@ func (r *JivaVolumeReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 		return reconcile.Result{}, err
 	}
 
+	if _, ok := podIPMap[instance.Name]; !ok {
+		err = r.updatePodIPMap(instance)
+		if err != nil {
+			// log err only, as controller must be in container creating state
+			// don't return err as it will dump stack trace unneccesary
+			logrus.Infof("failed to get controller pod ip for volume %s: %s", instance.Name, err.Error())
```
Comment (prateekpandey14) [inline pkg/controllers/jivavolume_controller.go:144]:
```suggestion
			logrus.Infof("not able to get controller pod ip for volume %s: %s", instance.Name, err.Error())
```

Comment (prateekpandey14) [reply pkg/controllers/jivavolume_controller.go:144]: not able is better word here since its expected during provisioning time.

**Comment by shubham14bajpai** [comment]
> Can you please verify the following scenarios where a node containing both controller and replica is brought down:
> 
> 1. Delete the controller pod, so that now 2 controller pods are present. One in deleting state and the other in Running/ContainerCreating state.
> 2. Verify if we are able to fetch the status even when one replica is down.

Tried the above scenario:
1. While there were two controller pods one in Running other in Terminating state the status was `Unknown`, the pod finally terminated after 3 minutes.
2. Once only a single controller pod was running it was able to read status of 2 replicas and eventually 3rd replica came.

File: pkg/controllers/jivavolume_controller.go (hunk 1 above)
Comment (shubham14bajpai) [reply pkg/controllers/jivavolume_controller.go:145]: To avoid log flooding with error msgs

File: pkg/controllers/jivavolume_controller.go [hunk 3]
```diff
@@ -179,6 +191,30 @@ func (r *JivaVolumeReconciler) Reconcile(ctx context.Context, req ctrl.Request)
...
+	err := r.List(context.TODO(), &pods, &client.ListOptions{
+		Namespace:     cr.Namespace,
+		LabelSelector: labelSelector,
+	})
+	if err != nil {
+		return err
+	}
+	if len(pods.Items) != 1 {
+		return fmt.Errorf("expected 1 controller pod got %d", len(pods.Items))
+	}
```
Comment (mittachaitu) [inline pkg/controllers/jivavolume_controller.go:212]: What if pods remains in evicted state due to draining of node?

**Review by mittachaitu — COMMENTED** [review]
provided few questions

File: pkg/controllers/jivavolume_controller.go (hunk 3 above)
Comment (shubham14bajpai) [reply pkg/controllers/jivavolume_controller.go]: Added check with fieldSelector

File: pkg/controllers/jivavolume_controller.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (shubham14bajpai) [reply pkg/controllers/jivavolume_controller.go]: This happens most;y when controller pod is in container creating state. Having it as info is good enough.

File: cmd/manager/main.go [hunk 4]
```diff
@@ -78,7 +78,7 @@ func main() {
 		os.Exit(1)
 	}
 
-	duration := 5 * time.Second
+	duration := 30 * time.Second
```
[the remark this answers is not in the document]
Comment (shubham14bajpai) [reply cmd/manager/main.go:81]: Will take this in another PR

**Comment by shubham14bajpai** [comment]
Observed another issue with replica movement on node deletion that the replica pod and pvc gets deleted again and again
```
$ k get pods -w
NAME                                                              READY   STATUS        RESTARTS   AGE
jiva-operator-99f6fbccb-72dkt                                     1/1     Running       0          3m24s
openebs-jiva-csi-controller-0                                     5/5     Running       0          111m
openebs-jiva-csi-node-lxtn2                                       3/3     Running       0          111m
openebs-jiva-csi-node-mq7qp                                       3/3     Running       0          74m
openebs-jiva-csi-node-mz9p6                                       3/3     Running       0          111m
openebs-localpv-provisioner-85485d7b48-9vknk                      1/1     Running       0          113m
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-ctrl-6c77fb5g27rm   2/2     Running       0          71s
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-ctrl-6c77fb5h8cms   2/2     Terminating   0          78m
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-rep-0               1/1     Running       1          74m
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-rep-1               1/1     Running       2          83m
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-rep-2               1/1     Running       2          83m
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-ctrl-6c77fb5h8cms   2/2     Terminating   0          80m
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-ctrl-6c77fb5h8cms   2/2     Terminating   0          80m
openebs-jiva-csi-node-lxtn2                                       3/3     Terminating   0          113m
openebs-jiva-csi-node-lxtn2                                       3/3     Terminating   0          113m
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-rep-1               1/1     Terminating   2          84m
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-rep-1               1/1     Terminating   2          84m
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-rep-1               0/1     Pending       0          0s
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-rep-1               0/1     Pending       0          0s
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-rep-1               0/1     Terminating   0          0s
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-rep-1               0/1     Terminating   0          0s
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-rep-1               0/1     Pending       0          0s
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-rep-1               0/1     Pending       0          0s
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-rep-1               0/1     Terminating   0          10s
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-rep-1               0/1     Terminating   0          10s
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-rep-1               0/1     Pending       0          0s
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-rep-1               0/1     Pending       0          0s
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-rep-1               0/1     Terminating   0          10s
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-rep-1               0/1     Terminating   0          10s
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-rep-1               0/1     Pending       0          0s
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-rep-1               0/1     Pending       0          0s
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-rep-1               0/1     Terminating   0          10s
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-rep-1               0/1     Terminating   0          10s
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-rep-1               0/1     Pending       0          0s
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-rep-1               0/1     Pending       0          0s
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-rep-1               0/1     Terminating   0          10s
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-rep-1               0/1     Terminating   0          10s
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-rep-1               0/1     Pending       0          0s
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-rep-1               0/1     Pending       0          0s
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-rep-1               0/1     Pending       0          0s
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-rep-1               0/1     Terminating   0          10s
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-rep-1               0/1     Terminating   0          10s
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-rep-1               0/1     Pending       0          0s
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-rep-1               0/1     Pending       0          0s
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-rep-1               0/1     Terminating   0          10s
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-rep-1               0/1     Terminating   0          10s
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-rep-1               0/1     Pending       0          0s
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-rep-1               0/1     Pending       0          0s
openebs-jiva-csi-node-rsgvt                                       0/3     Pending       0          0s
openebs-jiva-csi-node-rsgvt                                       0/3     Pending       0          0s
openebs-jiva-csi-node-rsgvt                                       0/3     ContainerCreating   0          0s
init-pvc-2d4799f5-24e4-4fd9-8bfd-81aa254f51b0                     0/1     Pending             0          0s
init-pvc-2d4799f5-24e4-4fd9-8bfd-81aa254f51b0                     0/1     Pending             0          0s
init-pvc-2d4799f5-24e4-4fd9-8bfd-81aa254f51b0                     0/1     ContainerCreating   0          0s
init-pvc-2d4799f5-24e4-4fd9-8bfd-81aa254f51b0                     0/1     Completed           0          3s
init-pvc-2d4799f5-24e4-4fd9-8bfd-81aa254f51b0                     0/1     Terminating         0          3s
init-pvc-2d4799f5-24e4-4fd9-8bfd-81aa254f51b0                     0/1     Terminating         0          3s
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-rep-1               0/1     Pending             0          7s
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-rep-1               0/1     ContainerCreating   0          7s
pvc-ea334577-d4dc-4eba-a4e3-d07bb6c33cd8-jiva-rep-1               1/1     Running             0          21s
openebs-jiva-csi-node-rsgvt                                       3/3     Running             
```

This was happening as the new pending pod does not have node name annotation on it. Added a check for that as well which fixes this issue.

**Review by prateekpandey14 — APPROVED** [review]
lgtm

**Review by mittachaitu — APPROVED** [review]
LGTM

**Pull request merged**